examples/tp/tm_high_order.py

```diff
@@ -0,0 +1,465 @@
+print "A predicts B, B predicts C, C predicts D, and D does not output any"
+print "prediction."
+print "N.B. Here, we are presenting individual characters, that is, a character"
+print "deprived of context in a sequence. There is no prediction for characters"
+print "X and Y as we have not presented them to the TM in any sequence."
+print "-"*50
+print ""
+
+showPredictions()
+
+print ""
+print "-"*50
+print "Part 2. We now present the sequence XBCY to the TM. As expected, the accuracy will"
+print "drop until the TM learns the new sequence (Fig 2). What will be the prediction of"
+print "the TM if presented with the sequence BC? This would depend on what character"
+print "anteceding B. This is an important feature of high-order sequences."
+print "-"*50
+print ""
+
+x = []
+y = []
+trainTM(seq2, timeSteps=10, noiseLevel=0.0)
+
+# In this figure you can see how the TM starts making good predictions for particular
+# characters (spikes in the plot). Then, it will get half of its predictions right, which
+# correspond to the times in which is presented with character C. After some time, it
+# will learn correctly the sequence XBCY, and predict its characters accordingly.
+plt.ylim([-0.1,1.1])
+plt.plot(x, y)
+plt.xlabel("Timestep")
+plt.ylabel("Prediction Accuracy")
+plt.title("Fig. 2: TM learns new sequence XBCY")
+plt.savefig("figure_2")
+plt.close()
+
+print ""
+print "-"*50
+print "We will present again each of the characters individually to the TM, that is,"
+print "not within any of the two sequences. When presented with character A the TM"
+print "predicts B, B predicts C, but this time C outputs a simultaneous prediction of"
+print "both D and Y. In order to disambiguate, the TM would require to know if the"
+print "preceding characters were AB or XB. When presented with character X the TM"
+print "predicts B, whereas Y and D yield no prediction."
+print "-"*50
+print ""
+
+showPredictions()
+
+# PART 3. Now we will present noisy inputs to the TM. We will add noise to the sequence XBCY
+# by corrupting 30% of its bits. We would like to see how the TM responds in the presence of
+# noise and how it recovers from it.
+print ""
+print "-"*50
+print "Part 3. We will add noise to the sequence XBCY by corrupting 30% of the bits in the vectors"
+print "encoding each character. We would expect to see a decrease in prediction accuracy as the"
+print "TM is unable to learn the random noise in the input (Fig 3). However, this decrease is not"
+print "significant."
+print "-"*50
+print ""
+
+x = []
+y = []
+trainTM(seq2, timeSteps=50, noiseLevel=0.3)
+
+plt.ylim([-0.1,1.1])
+plt.plot(x, y)
+plt.xlabel("Timestep")
+plt.ylabel("Prediction Accuracy")
+plt.title("Fig. 3: Accuracy in TM with 30% noise in input")
+plt.savefig("figure_3")
+plt.close()
+
+print ""
+print "-"*50
+print "Let's have a look again at the output of the TM when presented with noisy"
+print "input (30%). Here, the noise is low so the TM does not try to learn the noise,"
```

Comment:
The comment is inaccurate. The TM always learns unless you tell it not to, even when the noise is low. In this case it doesn't learn because you are explicitly setting learn to False in showPredictions.